draft: add disaster recovery guidance for NGINXaaS for Azure deployments (#493)
nginx/documentation: content/nginxaas-azure/quickstart/disaster-recovery.md

> ---

Reviewer: this does not need to be solved here but @JTorreG the list of quickstart documents is growing and DR does not really feel like a quickstart topic to me :)

Reviewer 2: I agree this not feel like a QS topic

Reviewer 3: DR isn't a quickstart. You can:

1. Rename Quickstart --> How-to
2. Put the DR doc at the root level in the left nav

In a separate PR, we'll improve the IA around use cases.

Author: I was on the fence as well but didn't want to create a new root document. I'll create that now since it seems like the best path forward.

Author: Moved to root. Resolving.

title: Disaster recovery
weight: 800
toc: true
url: /nginxaas/azure/quickstart/disaster-recovery/
type:
- how-to
---

> This guide describes how to configure disaster recovery (DR) for F5 NGINX as a Service for Azure deployments in separate Azure regions, ensuring upstream access remains available even if a region fails. The deployment architecture ensures users can access backend application servers (upstreams) continuously from an alternative region if the primary region becomes unavailable. The solution leverages Terraform, Azure Virtual Network (VNet) peering, and unique subnets to support failover.

Author: I don't think we should limit ourselves to saying that this guide is only applicable to paired regions. I can add `...deployments in separate (ideally [paired](https://learn.microsoft.com/en-us/azure/reliability/regions-paired)) Azure regions...`

> This guide describes how to configure disaster recovery (DR) for F5 NGINX as a Service for Azure deployments in separate Azure regions, ensuring upstream access remains available even if a region fails. The deployment architecture ensures users can access backend application servers (upstreams) continuously from an alternative region if the primary region becomes unavailable. The solution leverages Terraform, Azure Virtual Network (VNet) peering, and unique subnets to support failover.

Author: added language around paired regions. Resolving.

---

> **Architecture Overview**

Reviewer: Now that we have a diagram, you can see that the upstreams look odd when they are in just the one spot while not being in the second one. Uneven to the eyes if someone looks at it as well.

I think an improvement **for later on** can be a scenario where we recommend customers have their own upstreams running identically in the regions where they want such a scenarios. AKS is a good example of deploying the same upstream applications multiple times in different clusters that is a bit more modern and would make the scenario a bit nicer too.

Author: The document is showing them how 2 NGINXaaS deployments can share the upstreams. I think it'll be more confusing if each region had its own upstreams in the diagram. I think that was the ask by @amudukutore

Author: The architecture mimics the ask from @amudukutore. Resoving this.

**Architecture Overview**

```
+-------------------+         +-------------------+
|  Region 1         |         |  Region 2         |
|  VNet1            |         |  VNet2            |
|  +-------------+  | Peered  |  +-------------+  |
|  | Subnet A    |  |<------->|  | Subnet B    |  |
|  | NGINX #1    |  |         |  | NGINX #2    |  |
|  +-------------+  |         |  +-------------+  |
|  | Upstreams   |  |         |  | Upstreams   |  |

> |  +-------------+  |         |  +-------------+  |

Reviewer 4: This makes it seem like the upstreams are different.  Is that the case?

+-------------------+         +-------------------+

> ```

Author: I removed mermaid and added a lucidchart diagram. Resolving this. I've also added you as a collaborator on the lucid diagram if you'd like to edit anything.

> ```

Author: Resolving.

- Each region has its own VNet, subnet, and NGINXaaS for Azure deployment.

> - VNet peering enables cross-region connectivity.

Reviewer: it might make sense to emphasize more on the connectivity between the networks and why that is important here than the technology (VNET peering). We should still call it out but maybe make it an example. Something along the lines of:

```
Cross region connectivity to ensure that upstreams are reachable from either deployment. We will be using VNET peering in this guide to establish the connectivity.
```

Reviewer 5: I formatted @puneetsarna 's suggestion for consistency

```suggestion
- Cross region connectivity ensures that upstreams are reachable from either deployment. We use VNet peering in this guide to establish that connectivity.
```

Author: Added this. Resolving.

- Upstreams (e.g., VMs) are accessible from either NGINX deployment.

---

## Prerequisites

- Two Azure regions selected for DR.
- Unique, non-overlapping VNet and subnet address spaces for each region.
- Terraform 1.3+ and AzureRM provider 4.23+.

> **Note**: Each NGINX deployment **must run on separate subnets and non-overlapping address spaces**. This is critical for [Virtual Network (VNet) peering](https://learn.microsoft.com/en-us/azure/virtual-network/how-to-configure-subnet-peering) between the two regions. For example:
>
>  - Region 1 (Virtual Network - 1 Address Space): `10.0.0.0/16`

> >  - Region 2 (Virtual Network - 2 Address Space): `10.1.0.0/16`

Reviewer: nit even though these are examples:
- Try using smaller subnets just for better guidance as customers may not have a `/16`.
- Try using a different Class CIDR instead to avoid possibilities in misconfiguration if someone were to skim over this. Eg: `10.0.0.0/24` and `192.168.0.0/24`.

```suggestion
>  - Region 2 (Virtual Network - 2 Address Space): `192.168.1.0/24`
```

Reviewer 5: Oh.... is this `deployment_primary_subnet` and `deployment_secondary_subnet` as shown in "Step 1"? (or is it `primary_vnet` and `secondary_vnet`?)

Should you explain this so a reader can visualize which network they're using?

Author: @mjang This is the primary_vnet and secondary_vnet. I've made it clearer. Thank you
@puneetsarna I take your point about having subnets that someone can easily differentiate just by looking at the document. I'll change them to 10.0.0.0/16 and 172.16.0.0/16. However, it's easy to see the separation between the vnet and subnets if we stick to classical boundaries of /16 and /24. The only thing I'm trying to get across is that a vnet is obviously bigger than the subnet because it encompasses it. I don't believe that anyone would think that they need a /16 to implement this topology.

Author: I hope I've made it clearer now. Resolving and please reopen if you think otherwise.

---

> 
> 
> ## Step 1: Deploy Prerequisite Infrastructure

Reviewer 2: ```suggestion

## Configure disaster recovery 

### Step 1: Deploy prerequisite infrastructure
```

> Each region requires its own resource group, VNet, subnet, public IP, network security group, and user-assigned identity. Example allocation in the prerequisites module:

Reviewer: does each region require it's own RG? 🤔 

and why would you need user MIs to configure this? Or are you just calling out the best way of configuring each deployment?

Reviewer 5: Tell us more about the code that follows. From what I see, I can't tell if the code is already available somewhere in Azure, of if they have to add it to some configuration file. If you're asking the reader to add this to /etc/some/path/to/nginx.conf or maybe /etc/nginx/disaster_recovery.conf, say so.

Author: I've added all relevant code and redacted unimportant bits for brevity. I think we have a good balance now. Resoving this discussion.

```hcl
# Region 1

> resource "azurerm_virtual_network" "deployment_primary_vnet" {

Reviewer: for these VNETs:

At a minimum, you should:
- call out the regions in use in your configs as it is an important detail. As a suggestion, you could pick two paired Azure regions and use them as an example here.
- nit: maybe call out the RG as well since you are mentioning the need for different RGs?

Author: Done. Resolving.

  address_space = ["10.0.0.0/16"]
  # other config...
}
resource "azurerm_subnet" "deployment_primary_subnet" {
  address_prefixes = [cidrsubnet("10.0.0.0/16", 8, 0)] # results in 10.0.0.0/24
}

# Region 2
resource "azurerm_virtual_network" "deployment_secondary_vnet" {
  address_space = ["10.1.0.0/16"]
}
resource "azurerm_subnet" "deployment_secondary_subnet" {
  address_prefixes = [cidrsubnet("10.1.0.0/16", 8, 0)] # results in 10.1.0.0/24
}
```
---

> ## Step 2: Deploy NGINXaaS for Azure Deployment in Each Region

Reviewer 4: Nit - Step 1 and Step 2 have different heading sizes

Author: fixed, thanks!

```hcl
resource "azurerm_nginx_deployment" "deployment_primary_nginxaas" {
  name                = var.name_primary
  resource_group_name = var.resource_group_primary
  location            = var.location_primary
  ...
  network_interface {
    subnet_id = azurerm_subnet.deployment_primary_subnet.id
  }

> }

Reviewer: nit in all TF configs:

add a space in between resource blocks.

Author: Done. Resolving.

resource "azurerm_nginx_deployment" "deployment_secondary_nginxaas" {
  name                = var.name_secondary
  resource_group_name = var.resource_group_secondary
  location            = var.location_secondary
  ...
  network_interface {
    subnet_id = azurerm_subnet.deployment_secondary_subnet.id
  }
}
```

---

> ## Step 3: Peer the VNets

Reviewer 2: ```suggestion
### Step 3: Peer the VNets
```

```hcl
resource "azurerm_virtual_network_peering" "vnet_primary_to_vnet_secondary" {
  name                      = "peering-vnet-primary-to-vnet-secondary"
  resource_group_name       = var.resource_group_primary
  virtual_network_name      = azurerm_virtual_network.deployment_primary_vnet.name
  remote_virtual_network_id = azurerm_virtual_network.deployment_secondary_vnet.id
}
resource "azurerm_virtual_network_peering" "vnet_secondary_to_vnet_primary" {
  name                      = "peering-vnet-secondary-to-vnet-primary"
  resource_group_name       = var.resource_group_secondary
  virtual_network_name      = azurerm_virtual_network.deployment_secondary_vnet.name
  remote_virtual_network_id = azurerm_virtual_network.deployment_primary_vnet.id
}
```

---

> ## Step 4: Configure Upstreams

Reviewer 2: ```suggestion
### Step 4: Configure upstreams
```

> Deploy upstream VMs in a subnet separate from the NGINXaaS deployment subnet in the primary region. Example:

Reviewer: Since it's important to the setup, maybe highlight the primary region?
```suggestion
Deploy upstream VMs in a subnet separate from the NGINXaaS deployment subnet in the **primary region**. Example:
```

```hcl
resource "azurerm_subnet" "upstreams" {
  name                 = "upstreams-subnet"
  resource_group_name  = var.resource_group_primary
  virtual_network_name = azurerm_virtual_network.deployment_primary_vnet.name
  address_prefixes     = [cidrsubnet(var.vnet_addr_space, 8, 1)]
}
```

---

## Step 5: NGINX Configuration for Failover

> Configure both NGINX deployments to include upstreams from the primary regions in their load balancing configuration. Example `nginx.conf` snippet:

Author: Added this. Resolving.

```nginx
upstream backend {
    server <vm1-private-ip>:80;
    server <vm2-private-ip>:80;
}
```

- Use private IPs reachable via VNet peering.

> - Health checks can be configured to detect regional failures and reroute traffic.

Author: Done. Resolving.

> - Health checks can be configured to detect regional failures and reroute traffic.

Reviewer 4: Might want to expand on this (or a link to some external doc)

---

## Step 6: DNS and Failover

> - Use Azure Traffic Manager or an external DNS solution to direct traffic to the healthy NGINX deployment.

Author: Yes, I have. Resolving.

> - In case of a regional outage, update DNS to point to the public IP of the surviving region's NGINX.

Reviewer: ```suggestion
- In case of a regional outage, update DNS record to point to the public IP of the NGINXaas deployment in the secondary region.
```

---

## Workarounds for Subnet Peering Caveats

- **Subnet Peering for Overlapping VNets:**
If overlapping address spaces are unavoidable, use subnet-level peering to selectively peer only the required subnets.

```shell
az network vnet peering create \
  --name vnet1-to-vnet2 \
  --resource-group <rg1> \
  --vnet-name <vnet1> \
  --remote-vnet <vnet2-id> \
  --allow-vnet-access \
  --peer-complete-vnet false \
  --local-subnet-names <subnet1> \
  --remote-subnet-names <subnet2>
```

---

## Failover Process

1. **Monitor**: Continuously monitor NGINXaaS deployment health in both regions.
2. **Failover**: If a region fails, update DNS or Traffic Manager to route traffic to the surviving region's NGINXaaS deployment.
3. **Recovery**: Restore the failed region and verify peering and upstream connectivity before re-enabling traffic.

---

> ## Diagram

Reviewer: maybe move this diagram along side the other one you had above?

Author: Only one diagram now. Resolving.

> ```plaintext
> +-----------------------+      Peering      +-----------------------+
> |  Region 1             |<----------------->|  Region 2             |
> |  NGINXaaS Deployment 1|                   |  NGINXaaS Deployment 2|
> |  Upstreams (VMs)      |                   |                       |
> +-----------------------+                   +-----------------------+
>          |                                         |
>          +-------------------+   +-----------------+
>                              |   |
>                          Users/Clients (via DNS/Traffic Manager)
> ```

Reviewer 2: ````suggestion
```mermaid
flowchart LR

    subgraph Region1[Region 1]
        direction TB
        NGINX1[NGINXaaS Deployment 1]
        VM1["Upstreams (VMs)"]
    end

    subgraph Region2[Region 2]
        direction TB
        NGINX2[NGINXaaS Deployment 2]
    end

    Users["웃 Users/Clients<br/>&#40;using DNS/traffic manager&#41;"]

    Region2 <-->|Peering| Region1
    NGINX1 --> Users
    NGINX2 --> Users

    %% Styling
    style Region1 fill:#9bb1de,stroke:#4a90e2,stroke-width:2px
    style Region2 fill:#9bb1de,stroke:#4a90e2,stroke-width:2px
    style NGINX1 fill:#d9fade,stroke:#2e7d32,stroke-width:2px,color:#000
    style NGINX2 fill:#d9fade,stroke:#2e7d32,stroke-width:2px,color:#000
    style VM1 fill:#e8f3fe,stroke:#3075ff,stroke-width:2px,color:#000
    style Users fill:#faefd9,stroke:orange,color:orange,stroke-width:2px
```
````

Reviewer 2: ![image](https://github.com/user-attachments/assets/18224e12-4c1d-4c1e-8085-20c9304dc212)

Reviewer 2: ![image](https://github.com/user-attachments/assets/c30eab93-4757-4d84-a30c-45a83b74bc5a)

---

> **Summary:**

Reviewer: headers might be nicer.

> By deploying NGINX in separate regions with unique subnets and peered VNets, and configuring upstreams and DNS for failover, this topology ensures high availability and DR for your upstreams. Use subnet peering if address spaces overlap. Always monitor and test your failover paths.

Reviewer: ```suggestion
By deploying NGINXaaS in separate regions with unique subnets and peered VNets, and configuring upstreams and DNS for failover, this topology ensures high availability and DR for your applications. Use subnet peering if address spaces overlap. Lastly, always monitor and test your failover paths.
```

---

> title: Disaster Recovery

Reviewer 2: ```suggestion
title: Disaster recovery
```

weight: 800
toc: true
url: /nginxaas/azure/quickstart/disaster-recovery/
type:
- how-to
---

> ## Disaster Recovery Configuration for F5 NGINX as a Service for Azure Deployments

Reviewer 2: ```suggestion

```

Reviewer 2: We generally don't start docs with a heading that is an extended version of the title. I will look like a duplicate of the `title:` above in the page (almost same size and font)

![image](https://github.com/user-attachments/assets/7e803ea9-5b17-4bc8-9db1-92517e32208b)

>
>  - Region 1 (Virtual Network - 1 Address Space): `10.0.0.0/16`
>  - Region 2 (Virtual Network - 2 Address Space): `10.1.0.0/16`

---

## Configure disaster recovery 

### Step 1: Deploy prerequisite infrastructure

Each region requires its own resource group, VNet, subnet, public IP, network security group, and user-assigned identity. Example allocation in the prerequisites module:

```hcl
# Region 1
resource "azurerm_virtual_network" "deployment_primary_vnet" {
  address_space = ["10.0.0.0/16"]
  # other config...
}
resource "azurerm_subnet" "deployment_primary_subnet" {
  address_prefixes = [cidrsubnet("10.0.0.0/16", 8, 0)] # results in 10.0.0.0/24
}

# Region 2
resource "azurerm_virtual_network" "deployment_secondary_vnet" {
  address_space = ["10.1.0.0/16"]
}
resource "azurerm_subnet" "deployment_secondary_subnet" {
  address_prefixes = [cidrsubnet("10.1.0.0/16", 8, 0)] # results in 10.1.0.0/24
}
```
---

### Step 2: Deploy NGINXaaS for Azure in each region

> 

Reviewer 5: Can you summarize what the code shown in this step is doing? Something like: 

Add this code to /path/to/some.conf file. It supports deploying NGINXaaS for Azure in two supported regions. Do we know what's used for `location_primary` and `primary_subnet.id`? Does this come from the (#prerequisites)?

And you could say something like "In Step 3, you'll set up these virtual subnets as peers." (I presume that matches your reference to  2 different virtual networks (which you call VNets)

Author: Summarized. Resoving this. Please reopen if it's still unclear.

```hcl
resource "azurerm_nginx_deployment" "deployment_primary_nginxaas" {
  name                = var.name_primary
  resource_group_name = var.resource_group_primary
  location            = var.location_primary
  ...
  network_interface {
    subnet_id = azurerm_subnet.deployment_primary_subnet.id
  }
}
resource "azurerm_nginx_deployment" "deployment_secondary_nginxaas" {
  name                = var.name_secondary
  resource_group_name = var.resource_group_secondary
  location            = var.location_secondary
  ...
  network_interface {
    subnet_id = azurerm_subnet.deployment_secondary_subnet.id
  }
}
```

---

### Step 3: Peer the VNets

> 

Reviewer 5: I'm guessing that you're adding this code to the same file... it'd be nice to know which file.

Author: It doesn't matter with terraform because all things can go in just a single file. It's declarative. The thing for a reader of this doc to pay attention to would be the terraform resources used. I hope the new iteration of the doc makes things clearer. Please open a new discussion if things are still hazy. thanks

}
```

---

### Step 4: Configure upstreams

Deploy upstream VMs in a subnet separate from the NGINXaaS deployment subnet in the **primary region**. Example:

```hcl
resource "azurerm_subnet" "upstreams" {
  name                 = "upstreams-subnet"
  resource_group_name  = var.resource_group_primary
  virtual_network_name = azurerm_virtual_network.deployment_primary_vnet.name
  address_prefixes     = [cidrsubnet(var.vnet_addr_space, 8, 1)]
}
```

---

## Step 5: NGINX Configuration for Failover

Configure both NGINX deployments to include upstreams from the primary regions in their load balancing configuration. Example `nginx.conf` snippet:

```nginx
upstream backend {
    server <vm1-private-ip>:80;
    server <vm2-private-ip>:80;
}
```

- Use private IPs reachable via VNet peering.
- Health checks can be configured to detect regional failures and reroute traffic.

---

### Step 6: DNS and failover

- Use Azure Traffic Manager or an external DNS solution to direct traffic to the healthy NGINX deployment.
- In case of a regional outage, update DNS record to point to the public IP of the NGINXaas deployment in the secondary region.

---

> ## Workarounds for Subnet Peering Caveats

Reviewer 5: ```suggestion
## Workarounds for Subnet Peering
```

Author: Removed this section. Resolving.
